### ISSUE TITLE
Prefer composition over multiple inheritance

### ISSUE THREAD

**Issue by karlhigley** [OpenMined/PySyft#2894]
**Is your feature request related to a problem? Please describe.**
Understanding and debugging classes with multiple inheritance is rather difficult. A lot of the behavior is implicit instead of explicit, which isn't very Pythonic and makes it hard to follow.

**Describe the solution you'd like**
Refactor classes with multiple inheritance to convert one of the parents from an is-a relationship to a has-a relationship. For example, "a worker has object storage", instead of "a worker is object storage."

**Describe alternatives you've considered**
- Honestly can't think of a reasonable alternative

**Additional context**
Multiply inheriting PySyft classes:
* [x] `Promise`
* [ ] `PointerTensor`
* [x] `Plan`
* [x] `BaseWorker`
* [x] `VirtualWorker`
* [x] `WebSocketServerWorker`

**Comment by bhadreshpsavani** [comment]
Can i work on this issue?

**Comment by karlhigley** [comment]
@bhadreshpsavani Assigned it to you. Happy to help with this, ping me on Slack if you want to talk more about it.

**Comment by bhadreshpsavani** [comment]
sure @karlhigley

**Comment by karlhigley** [comment]
`Plan` no longer has multiple parents and `Promise` has been removed.

**Comment by karlhigley** [comment]
@tudorcebere This might be helpful for sorting out some of the worker issues you've been looking at.

**Comment by karlhigley** [comment]
At this point, I think a series of recent refactors have addressed all of the listed classes except `PointerTensor`. I don't have a clear idea of how to make `PointerTensor` singly inheriting—I think it needs to have the `AbstractTensor` API, so maybe the better option is to make it be an `AbstactTensor` and have an `ObjectPointer`. What do you think, @Jasopaum?

**Comment by Jasopaum** [comment]
I think `PointerTensor` is one of the cases where I'd be kind of in favor of the multiple inheritance.
Maybe some of the methods on `PointerTensor` should be moved to `ObjectPointer` though.
Then `PointerTensor` would have a very simple implementation with almost no method of its own.
But I'm aware that multiple inheritance complexifies some things so...

**Comment by karlhigley** [comment]
I'm good with that, and I think we have some `ObjectPointer` refactoring to do anyway, so let's close this one and call it done. 👍
